File: Ios/BarkCloud/BarkCloud/Features/Vault/VaultViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class VaultViewModel {
    enum LockState { case locked, unlocking, unlocked }

    var lockState: LockState = .locked
    var isSelecting = false
    var selection: Set<String> = []

    private let vault: VaultStore
    private let biometric: BiometricGate

    init(vault: VaultStore, biometric: BiometricGate) {
        self.vault = vault
        self.biometric = biometric
    }

    var items: [VaultItem] { vault.items }
    var isEmpty: Bool { vault.isEmpty }

    /// Запросить биометрию и открыть сейф при успехе.
    func unlock() async {
        guard lockState != .unlocked else { return }
        lockState = .unlocking
        let ok = await biometric.authenticate(reason: String(localized: "vault_unlock_reason"))
        lockState = ok ? .unlocked : .locked
    }

    /// Снова запереть (уход в фон / выход с экрана).
    func lock() {
        lockState = .locked
        exitSelection()
    }

    // MARK: - Мультивыбор «убрать из сейфа»

    func enterSelection() { isSelecting = true }

    func exitSelection() {
        isSelecting = false
        selection = []
    }

    func toggle(_ id: String) {
        if selection.contains(id) { selection.remove(id) } else { selection.insert(id) }
    }

    var hasSelection: Bool { !selection.isEmpty }

    /// Убрать выбранные элементы из сейфа — они вернутся в обычную галерею.
    func removeSelected() {
        vault.remove(ids: selection)
        exitSelection()
    }
}

File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/ui/vault/VaultViewModel.kt
package com.barkfluff.BarkCloud.ui.vault

import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.barkfluff.BarkCloud.BarkCloudApplication
import com.barkfluff.BarkCloud.data.BiometricGate
import com.barkfluff.BarkCloud.data.vault.VaultItem
import com.barkfluff.BarkCloud.data.vault.VaultStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class VaultLockState { LOCKED, UNLOCKING, UNLOCKED }

data class VaultUiState(
    val lockState: VaultLockState = VaultLockState.LOCKED,
    val items: List<VaultItem> = emptyList(),
)

/**
 * Per-session блокировка (без grace-period — проще, чем App Lock): релочится
 * при каждом уходе экрана из foreground ([relock], вызывается из [VaultScreen]
 * на ON_STOP), зеркалит iOS scenePhase-поведение VaultViewModel.
 */
class VaultViewModel(private val store: VaultStore) : ViewModel() {

    private val _state = MutableStateFlow(VaultUiState())
    val state: StateFlow<VaultUiState> = _state.asStateFlow()

    fun unlock(activity: FragmentActivity, title: String) {
        if (_state.value.lockState != VaultLockState.LOCKED) return
        _state.update { it.copy(lockState = VaultLockState.UNLOCKING) }
        viewModelScope.launch {
            val ok = BiometricGate.authenticate(activity, title)
            _state.update {
                if (ok) it.copy(lockState = VaultLockState.UNLOCKED, items = store.items())
                else it.copy(lockState = VaultLockState.LOCKED)
            }
        }
    }

    fun relock() {
        _state.update { it.copy(lockState = VaultLockState.LOCKED) }
    }

    fun remove(fileId: String) {
        store.remove(fileId)
        _state.update { it.copy(items = it.items.filterNot { item -> item.fileId == fileId }) }
    }

    companion object {
        fun factory(): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]
                    as BarkCloudApplication
                return VaultViewModel(app.vaultStore) as T
            }
        }
    }
}
